src/DynamoManipulation/Gizmo.cs

```diff
@@ -0,0 +1,174 @@
+        /// <param name="direction">View projection direction.</param>
+        /// <param name="hitObject">The object hit by the input ray.</param>
+        /// <returns>True when an object is hit, else false.</returns>
+        bool HitTest(Point source, Vector direction, out object hitObject);
+
+        /// <summary>
+        /// Computes new offset based on the hit point.
+        /// </summary>
+        /// <param name="newPosition">New position of mouse click.</param>
+        /// <param name="viewDirection">view projection direction</param>
+        /// <returns>Offset vector of hit point wrt it's origin.</returns>
+        Vector GetOffset(Point newPosition, Vector viewDirection);
+
+        /// <summary>
+        /// Gets render package for all the drawables of this Gizmo.
+        /// </summary>
+        /// <returns>List of render packages.</returns>
+        IEnumerable<IRenderPackage> GetDrawables();
+
+        /// <summary>
+        /// 
+        /// </summary>
+        void DeleteTransientGraphics();
+
+        /// <summary>
+        /// Highlight gizmo drawables or create transient geometry to highlight the gizmo during mouse over
+        /// </summary>
+        /// <returns></returns>
+        void HighlightGizmo();
+
+        /// <summary>
+        /// Unhighlight gizmo drawables or delete all transient geometry used to highlight gizmo during mouse over 
+        /// </summary>
+        void UnhighlightGizmo();
+
+        /// <summary>
+        /// 
+        /// </summary>
+        void UpdateGizmoGraphics();
+    }
+
+    internal abstract class Gizmo : IGizmo
+    {
+        private const double zDepth = 20.0;
+
+        private readonly NodeManipulator manipulator;
+
+        private Point3D? cameraPosition;
+
+        protected IWatch3DViewModel BackgroundPreviewViewModel
+        {
+            get { return manipulator.BackgroundPreviewViewModel; }
+        }
+
+        protected IRenderPackageFactory RenderPackageFactory
+        {
+            get { return manipulator.RenderPackageFactory; }
+        }
+
+        protected Point PointOrigin
+        {
+            get { return manipulator.Origin; }
+        }
+
+        private string name = "gizmo";
+        public string Name 
+        { 
+            get { return name; }
+            set
+            {
+                if (!name.Contains(value))
+                {
+                    name = string.Format("{0}_{1}", name, value);
+                }
+            } 
+        }
+
+        public Point Origin
+        {
+            get
+            {
+                var cameraPos = cameraPosition != null ?
+                    Point.ByCoordinates(cameraPosition.Value.X, cameraPosition.Value.Y, cameraPosition.Value.Z) : null;
+
+                if (cameraPos == null) throw new Exception("camera position is null");
+
+                var vec = Vector.ByTwoPoints(cameraPos, PointOrigin).Normalized();
+                return cameraPos.Add(vec.Scale(zDepth));
```

Comment:
Everytime the gizmo origin is queried it is computed based on the current point position and current camera location.